File: genai_telemetry/exporters/elasticsearch.py
"""
Elasticsearch exporter.
"""

import atexit
import base64
import json
import logging
import ssl
import threading
import time
import urllib.request
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from genai_telemetry.exporters.base import BaseExporter

logger = logging.getLogger("genai_telemetry.exporters.elasticsearch")


class ElasticsearchExporter(BaseExporter):
    """Sends spans to Elasticsearch."""
    
    def __init__(
        self,
        hosts: List[str] = None,
        index: str = "genai-traces",
        api_key: str = None,
        username: str = None,
        password: str = None,
        verify_ssl: bool = True,
        batch_size: int = 1,
        flush_interval: float = 5.0
    ):
        """
        Initialize Elasticsearch exporter.
        
        Args:
            hosts: List of Elasticsearch host URLs
            index: Index name prefix (date will be appended)
            api_key: Elasticsearch API key for authentication
            username: Username for basic auth
            password: Password for basic auth
            verify_ssl: Whether to verify SSL certificates
            batch_size: Number of events to batch before sending
            flush_interval: Seconds between automatic flushes
        """
        self.hosts = hosts or ["http://localhost:9200"]
        self.index = index
        self.api_key = api_key
        self.username = username
        self.password = password
        self.batch_size = batch_size
        self.flush_interval = flush_interval
        
        self.ssl_context = ssl.create_default_context()
        if not verify_ssl:
            self.ssl_context.check_hostname = False
            self.ssl_context.verify_mode = ssl.CERT_NONE
        
        self._batch: List[dict] = []
        self._lock = threading.Lock()
        self._flush_thread: Optional[threading.Thread] = None
        self._running = False
        self._host_index = 0
        
        atexit.register(self.stop)
    
    def _get_host(self) -> str:
        """Round-robin host selection."""
        host = self.hosts[self._host_index % len(self.hosts)]
        self._host_index += 1
        return host.rstrip("/")
    
    def _get_headers(self) -> dict:
        """Build authentication headers."""
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"ApiKey {self.api_key}"
        elif self.username and self.password:
            credentials = base64.b64encode(f"{self.username}:{self.password}".encode()).decode()
            headers["Authorization"] = f"Basic {credentials}"
        return headers
    
    def start(self) -> None:
        """Start the exporter and background flush thread."""
        if self._running:
            return
        self._running = True
        if self.batch_size > 1:
            self._flush_thread = threading.Thread(target=self._flush_loop, daemon=True)
            self._flush_thread.start()
    
    def stop(self) -> None:
        """Stop the exporter and flush remaining data."""
        self._running = False
        self.flush()
    
    def _flush_loop(self) -> None:
        """Background thread for periodic flushing."""
        while self._running:
            time.sleep(self.flush_interval)
            self.flush()
    
    def flush(self) -> None:
        """Flush any buffered spans."""
        with self._lock:
            if not self._batch:
                return
            batch = self._batch.copy()
            self._batch = []
        self._send_batch(batch)
    
    def _send_batch(self, batch: List[dict]) -> bool:
        """Send a batch using Elasticsearch bulk API."""
        if not batch:
            return True
        
        # Build bulk request
        lines = []
        for span in batch:
            # Index action
            index_date = datetime.now().strftime("%Y.%m.%d")
            index_name = f"{self.index}-{index_date}"
            action = {"index": {"_index": index_name}}
            lines.append(json.dumps(action))
            lines.append(json.dumps(span))
        
        payload = "\n".join(lines) + "\n"
        return self._send(payload, "/_bulk")
    
    def _send(self, payload: str, endpoint: str = "") -> bool:
        """Send payload to Elasticsearch."""
        host = self._get_host()
        url = f"{host}{endpoint}"
        data = payload.encode("utf-8")
        
        req = urllib.request.Request(
            url,
            data=data,
            headers=self._get_headers(),
            method="POST"
        )
        
        try:
            ctx = self.ssl_context if url.startswith("https") else None
            with urllib.request.urlopen(req, context=ctx, timeout=10) as resp:
                return resp.status in (200, 201)
        except Exception as e:
            logger.error(f"Elasticsearch Error: {e}")
            return False
    
    def export(self, span_data: Dict[str, Any]) -> bool:
        """Export a single span."""
        # Add @timestamp for Elasticsearch
        if "@timestamp" not in span_data:
            span_data["@timestamp"] = span_data.get("timestamp", datetime.now(timezone.utc).isoformat())
        
        if self.batch_size <= 1:
            return self._send_batch([span_data])
        
        with self._lock:
            self._batch.append(span_data)
            should_flush = len(self._batch) >= self.batch_size
        
        if should_flush:
            self.flush()
        return True
    
    def health_check(self) -> bool:
        """Check if Elasticsearch is reachable."""
        try:
            host = self._get_host()
            req = urllib.request.Request(f"{host}/_cluster/health", headers=self._get_headers())
            with urllib.request.urlopen(req, timeout=5) as resp:
                return resp.status == 200
        except:
            return False

File: js/src/exporters/elasticsearch.ts
/**
 * Elasticsearch exporter.
 */

import type { SpanData } from '../types';
import { BaseExporter } from './base';

export interface ElasticsearchExporterOptions {
  hosts: string[];
  index?: string;
  apiKey?: string;
  username?: string;
  password?: string;
  verifySsl?: boolean;
  batchSize?: number;
  flushInterval?: number;
}

export class ElasticsearchExporter extends BaseExporter {
  private hosts: string[];
  private index: string;
  private apiKey?: string;
  private username?: string;
  private password?: string;
  private batchSize: number;
  private flushInterval: number;
  
  private batch: SpanData[] = [];
  private flushTimer?: ReturnType<typeof setInterval>;
  private running: boolean = false;
  private hostIndex: number = 0;

  constructor(options: ElasticsearchExporterOptions) {
    super();
    this.hosts = options.hosts.map(h => h.replace(/\/$/, ''));
    this.index = options.index || 'genai-traces';
    this.apiKey = options.apiKey;
    this.username = options.username;
    this.password = options.password;
    this.batchSize = options.batchSize || 1;
    this.flushInterval = (options.flushInterval || 5) * 1000;
  }

  private getHost(): string {
    const host = this.hosts[this.hostIndex];
    this.hostIndex = (this.hostIndex + 1) % this.hosts.length;
    return host;
  }

  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (this.apiKey) {
      headers['Authorization'] = `ApiKey ${this.apiKey}`;
    } else if (this.username && this.password) {
      const auth = Buffer.from(`${this.username}:${this.password}`).toString('base64');
      headers['Authorization'] = `Basic ${auth}`;
    }

    return headers;
  }

  start(): void {
    if (this.running) return;
    this.running = true;

    if (this.batchSize > 1) {
      this.flushTimer = setInterval(() => this.flush(), this.flushInterval);
    }
  }

  stop(): void {
    this.running = false;
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = undefined;
    }
    this.flush();
  }

  async flush(): Promise<void> {
    if (this.batch.length === 0) return;

    const batchToSend = [...this.batch];
    this.batch = [];

    await this.sendBatch(batchToSend);
  }

  private async sendBatch(spans: SpanData[]): Promise<boolean> {
    if (spans.length === 0) return true;

    // Build bulk request body
    const lines: string[] = [];
    for (const span of spans) {
      lines.push(JSON.stringify({ index: { _index: this.index } }));
      lines.push(JSON.stringify(span));
    }
    const payload = lines.join('\n') + '\n';

    try {
      const host = this.getHost();
      const response = await fetch(`${host}/_bulk`, {
        method: 'POST',
        headers: this.getHeaders(),
        body: payload,
      });

      if (!response.ok) {
        console.error('Elasticsearch error:', await response.text());
        return false;
      }

      const result = await response.json() as { errors?: boolean };
      return !result.errors;
    } catch (error) {
      console.error('Elasticsearch Error:', error);
      return false;
    }
  }

  async export(spanData: SpanData): Promise<boolean> {
    if (this.batchSize <= 1) {
      return this.sendBatch([spanData]);
    }

    this.batch.push(spanData);

    if (this.batch.length >= this.batchSize) {
      await this.flush();
    }

    return true;
  }

  async healthCheck(): Promise<boolean> {
    try {
      const host = this.getHost();
      const response = await fetch(`${host}/_cluster/health`, {
        headers: this.getHeaders(),
      });
      return response.ok;
    } catch {
      return false;
    }
  }
}
